lib/vnet/ssh_handler.go

```diff
@@ -61,13 +81,93 @@ func (h *sshHandler) handleTCPConnectorWithTargetConn(
 	connector func() (net.Conn, error),
 	targetConn net.Conn,
 ) error {
-	// For now we accept the incoming TCP conn to indicate that the node exists,
-	// but SSH connection forwarding is not implemented yet so we immediately
-	// close it.
+	hostCert, err := h.getHostCert(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
 	localConn, err := connector()
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	localConn.Close()
+	defer localConn.Close()
+
+	// For now we accept the incoming SSH connection but forwarding to the
+	// target is not implemented yet so we immediately close it.
+	serverConfig := &ssh.ServerConfig{
+		PublicKeyCallback: func(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+			if !sshutils.KeysEqual(h.cfg.sshProvider.trustedUserPublicKey, key) {
+				return nil, trace.AccessDenied("SSH client public key is not trusted")
+			}
+			return &ssh.Permissions{}, nil
+		},
+	}
+	serverConfig.AddHostKey(hostCert)
+	serverConn, chans, reqs, err := ssh.NewServerConn(localConn, serverConfig)
+	if err != nil {
+		return trace.Wrap(err, "accepting incoming SSH connection")
+	}
+	// Immediately close the connection but make sure to drain the channels.
+	serverConn.Close()
+	go ssh.DiscardRequests(reqs)
+	go func() {
+		for newChan := range chans {
+			_ = newChan.Reject(0, "")
+		}
+	}()
+	target := h.cfg.target
+	log.DebugContext(ctx, "Accepted incoming SSH connection",
+		"profile", target.profile,
+		"cluster", target.cluster,
+		"host", target.host,
+		"user", serverConn.User(),
+	)
 	return trace.NotImplemented("VNet SSH connection forwarding is not yet implemented")
 }
+
+func (h *sshHandler) getHostCert(ctx context.Context) (ssh.Signer, error) {
+	h.hostCertMu.Lock()
+	defer h.hostCertMu.Unlock()
+
+	// If the current host cert is not expiring in the next minute, return it.
+	now := h.cfg.clock.Now()
+	if now.Before(h.hostCertExpires.Add(-time.Minute)) {
+		return h.hostCert, nil
+	}
+
+	// Generate a new host cert.
+	hostCert, err := newSelfSignedHostCert(h.cfg.target.fqdn, h.cfg.sshProvider.hostCASigner, now)
+	if err != nil {
+		return nil, trace.Wrap(err, "generating host certificate")
+	}
+	log.DebugContext(ctx, "Generated new SSH host cert", "fqdn", h.cfg.target.fqdn)
+	h.hostCert = hostCert
+	h.hostCertExpires = now.Add(hostCertTTL)
+	return hostCert, nil
+}
+
+func newSelfSignedHostCert(hostFQDN string, signer ssh.Signer, now time.Time) (ssh.Signer, error) {
+	// The cert should be valid for the domain the user connected to which may
+	// or may not be fully-qualified with a trailing "."
```

Comment:
Wait, what? Should we be worried for the rest of Teleport?

Reply:
i was probably overzealous here in trying to support fully-qualified domains, nobody actually types in the trailing dot do they? Most of the SSH protocol treats "example.com" as completely different from "example.com.", and I don't think really any of our other SSH certs include the fully-qualified variant